### PR TITLE
Order resources for apply

### PR DESCRIPTION
**Pull request by phanimarupaka**
*Why*

This PR is to order the kubernetes resources before applying to ensure smooth creation without dependency conflicts.

### CODE REVIEW

File: pkg/apply/ResourceInfos_test.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		"apiVersion": "v1",
+		"kind":       "Deployment",
+		"metadata": map[string]interface{}{
+			"name":      "testdeployment",
+			"namespace": "testspace",
+		},
+	},
+}
+
+func TestResourceOrdering(t *testing.T) {
```
Comment (phanimarupaka) [inline pkg/apply/ResourceInfos_test.go]: Will add more such tests for more resource kinds if needed.

File: pkg/apply/ResourceInfos.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2020 The Kubernetes Authors.
```
Comment (mortent) [inline pkg/apply/ResourceInfos.go]: Files should be snake case, not camel case.

File: pkg/apply/ResourceInfos.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+func (a ResourceInfos) Less(i, j int) bool {
+	if !getGvk(a[i]).Equals(getGvk(a[j])) {
+		return getGvk(a[i]).IsLessThan(getGvk(a[j]))
+	}
+	return a[i].String() < a[j].String()
+}
+
+func getGvk(info *resource.Info) Gvk {
+	gvk := info.Object.GetObjectKind().GroupVersionKind()
+	return Gvk{
```
Comment (mortent) [inline pkg/apply/ResourceInfos.go]: Why are we creating a new type for GVK here? It seems unnecessary.

File: pkg/apply/ResourceInfos_test.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		Name:   "testdeployment",
+		Object: &deploymentObj,
+	}
+
+	infos := []*resource.Info{&deploymentInfo, &configMapInfo, &namespaceInfo}
+	sort.Sort(ResourceInfos(infos))
+
+	var result string
+	for _, info := range infos {
+		result += fmt.Sprintf("Name: %s, Kind: %s\n", info.Name, info.Object.GetObjectKind().GroupVersionKind().Kind)
```
Comment (mortent) [inline pkg/apply/ResourceInfos_test.go]: Can we find a better way to do the comparison here? Printing it and then comparing strings seems a bit strange.

File: pkg/apply/ResourceInfos_test.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+	expected := `Name: testspace, Kind: Namespace
+Name: the-map, Kind: ConfigMap
+Name: testdeployment, Kind: Deployment
+`
+
+	assert.Equal(t, result, expected)
+}
+
+func TestGvkLessThan(t *testing.T) {
```
Comment (mortent) [inline pkg/apply/ResourceInfos_test.go]: Turn this into table tests.

File: pkg/apply/gvk.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
+// Copyright 2020 The Kubernetes Authors.
```
Comment (mortent) [inline pkg/apply/gvk.go]: This file seems to be copied from Kustomize. Aside from the list of resources to decide how we should sort, I'm not sure what we actually need.

File: pkg/apply/ResourceInfos.go (hunk 2 above)
Comment (phanimarupaka) [reply pkg/apply/ResourceInfos.go]: Will do

File: pkg/apply/ResourceInfos.go (hunk 3 above)
Comment (phanimarupaka) [reply pkg/apply/ResourceInfos.go]: Will do

File: pkg/apply/ResourceInfos_test.go (hunk 4 above)
Comment (phanimarupaka) [reply pkg/apply/ResourceInfos_test.go]: I see that Sprintf doesn't print but just forms and returns the string. Am I missing something.

// Sprintf formats according to a format specifier and returns the resulting string.

File: pkg/apply/ResourceInfos_test.go (hunk 5 above)
Comment (phanimarupaka) [reply pkg/apply/ResourceInfos_test.go]: This is only single test and is kinda trivial. Do we need table tests for that ?

File: pkg/apply/gvk.go (hunk 6 above)
Comment (phanimarupaka) [reply pkg/apply/gvk.go]: Will do.

File: pkg/apply/resource_infos.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	"PriorityClass",
+	"Deployment",
+	"StatefulSet",
+	"CronJob",
+	"PodDisruptionBudget",
+}
+var orderLast = []string{
+	"ValidatingWebhookConfiguration",
+}
+var typeOrders = func() map[string]int {
```
Comment (mortent) [inline pkg/apply/resource_infos.go]: This might be more readable with a named function?

File: pkg/apply/ResourceInfos_test.go (hunk 4 above)
Comment (mortent) [reply pkg/apply/ResourceInfos_test.go]: Yeah, so it doesn't print to Stdout. I meant that turning the results into strings and the comparing it seems strange. Can we just verify that the order of the types follow what is expected?

File: pkg/apply/ResourceInfos_test.go (hunk 5 above)
Comment (mortent) [reply pkg/apply/ResourceInfos_test.go]: yeah, if this test is sufficient for testing the function, then it is ok.

File: pkg/apply/resource_infos.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+func (a ResourceInfos) Len() int      { return len(a) }
+func (a ResourceInfos) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ResourceInfos) Less(i, j int) bool {
+	x := a[i].Object.GetObjectKind().GroupVersionKind()
+	o := a[j].Object.GetObjectKind().GroupVersionKind()
+	if !Equals(x, o) {
+		return IsLessThan(x, o)
+	}
+	return a[i].String() < a[j].String()
```
Comment (mortent) [inline pkg/apply/resource_infos.go]: So `Info` contains a lot of fields, and it is not obvious what we are doing when comparing the `String` functions. Also, we end up depending on the implementation of `String`, which could easily change. I suggest we should be more explicit about what we are doing here.

Comment (phanimarupaka) [reply pkg/apply/resource_infos.go]: This is just a tie-breaker and if this line is being executed, the result shouldn't matter as both options are same and their ordering doesn't hurt. This is how it is done in kustomize currently and I borrowed the logic to be consistent.

File: pkg/apply/resource_infos.go (hunk 7 above)
Comment (phanimarupaka) [reply pkg/apply/resource_infos.go]: Done

**Comment by mortent** [comment]
Failing build is due to the known out-of-memory issue on linux with Travis. I'm merging this manually.

**Pull request merged**